News feeds as page controller

bezin opened the pull request:
This implements news feeds as page controller and superseds #4459 eventually. Also I decided to split these into a news-bundle PR and an Event-bundle PR, because I find it easier to handle. 

Still a couple things to do: 
* [x] Add feeds to layout (DCA is updated, but the correct routes need to be added to the page. Requires `ResponseContext`)
* [x] Check and improve extensibility (e.g. https://github.com/contao/core-bundle/issues/1243, overwrite image size of every image element)
* [x] Get rid of some legacy code. if possible
* [x] Update permisssion handling (see below)
* [x] Page Icon: fix offset in page tree and add required variants (not visible, not editable, ...)
* [x] Update News Feed Insert tag
* [x] Remove obsolete code
* [x] Write migration (migrate feeds to pages; update permissions, if necessary)
* [x] Add tests

There are also some permissions for news feeds, i.e. which news feed a editor has access to and if he/she is allowed to create or delete them). Since this feed is now a page controller, access in general  is handled by the page type limitation. The question is, if we want to keep this distinction? I for myself are fine with the general limitation (can access news feed page types yes/no). 

# New Features
* JSON Feeds
* Only list featured/unfeatured news articles in the feed (equivalent to the newslist module)

aschempp commented on news-bundle/src/ContaoManager/Plugin.php:
```diff
@@ -12,23 +12,28 @@
+
+    public function registerContainerConfiguration(LoaderInterface $loader, array $managerConfig): void
+    {
+        $loader->load('@DebrilRssAtomBundle/Resources/config/services.yml');
```
this looks like a leftover, I thought we're not using the bundle anymore?

aschempp commented on news-bundle/src/Controller/NewsFeedController.php:
```diff
@@ -0,0 +1,257 @@
+    {
+        $feed = new Feed();
+        $feed->setTitle($pageModel->title);
+        $feed->setDescription($pageModel->feedDescription);
```
any particular reason we cannot use the existing (meta) description field for that?

bezin replied:
Nope, not really, have updated accordingly

aschempp commented on news-bundle/src/Controller/NewsFeedController.php:
```diff
@@ -0,0 +1,257 @@
+        $time = time();
+        $lastModified = null;
+
+        $archives = StringUtil::deserialize($pageModel->newsArchives);
```
Might want to add a check that this results in a non-empty array and return a 500 or 204 otherwise?

aschempp commented on news-bundle/src/Controller/NewsFeedController.php:
```diff
@@ -0,0 +1,257 @@
+
+        foreach ($articles as $article) {
+            // Never add unpublished elements to the RSS feeds
+            if (!$article->published || ($article->start && $article->start > $time) || ($article->stop && $article->stop <= $time)) {
```
I think this check was added for when the XML file was written. Now that we have a real controller, we can actually output unpublished items if front end preview is enabled, since the feed will only be for this user 😎

aschempp commented on news-bundle/src/Controller/NewsFeedController.php:
```diff
@@ -0,0 +1,257 @@
+
+            if ($lastModified < $item->getLastModified()) {
+                $lastModified = $item->getLastModified();
+            }
```
```suggestion
            $lastModified = null === $lastModified ? $item->getLastModified() : min($lastModified, $item->getLastModified());
```

aschempp commented on news-bundle/src/Controller/NewsFeedController.php:
```diff
@@ -0,0 +1,257 @@
+        $response->headers->set('Content-Type', $this->contentTypes[$pageModel->feedFormat]);
+
+        foreach ($archives as $archive) {
+            $this->responseTagger->addTags(['contao.db.news_archive.'.$archive]);
```
```suggestion
            $this->responseTagger->addTags(['contao.db.tl_news_archive.'.$archive]);
```

aschempp commented on news-bundle/src/Controller/NewsFeedController.php:
```diff
@@ -0,0 +1,257 @@
+        'rss' => 'application/rss+xml',
+    ];
+
+    public function __construct(private ContaoFramework $framework, private FeedIo $feedIo, private ResponseTagger $responseTagger, private ContaoContext $contaoContext, private ImageFactory $imageFactory, private RequestStack $requestStack, private InsertTagParser $insertTags, private string $projectDir)
```
`$responseTagger` service must be optional since it only exists with a reverse proxy. Also need to add respective checks when calling the `->addTags()` method

aschempp commented on news-bundle/src/Controller/NewsFeedController.php:
```diff
@@ -0,0 +1,257 @@
+        /** @var UserModel $authorModel */
+        if (($authorModel = $model->getRelated('author')) instanceof UserModel) {
+            $item->setAuthor((new Author())->setName($authorModel->name));
+        }
```
has this already been there? Not sure I like having a backend user name on news items in general, but dunno if a configuration makes sense 🤔

bezin replied:
No, it has not, but to me this belongs here to be complete. It is also [already done for the schema.org data](https://github.com/contao/contao/blob/5.x/news-bundle/src/Resources/contao/classes/News.php#L392-L399) and it is not the user name but the regular name.

aschempp replied:
true. But for schema.org data (and news templates) I can change it through the template. 🤔

aschempp commented on news-bundle/src/Controller/NewsFeedController.php:
```diff
@@ -0,0 +1,257 @@
+    {
+        // TODO: Is this still necessary?
+        // Push a new request to the request stack (#3856)
+        $request = Request::create($item->getLink());
```
this is no longer necessary for sure

aschempp commented on news-bundle/src/Controller/NewsFeedController.php:
```diff
@@ -0,0 +1,257 @@
+        return array_unique(array_values($this->urlSuffixes));
+    }
+
+    private function transformModelToItem(NewsModel $model, string $source, PageModel $pageModel): Item
```
maybe call it `$feedSource` instead of `$source` to be consistent with page model and other methods

aschempp commented on news-bundle/src/EventListener/DataContainer/LayoutListener.php:
```diff
@@ -0,0 +1,48 @@
+        if (null !== $feeds) {
+            while ($feeds->next()) {
+                $label = $this->translator->trans('FEED_FORMAT.'.$feeds->feedFormat, [], 'contao_tl_page');
+                $options[$feeds->id] = $feeds->title.sprintf(' (%s)', $label);
```
```suggestion
                $options[$feeds->id] = sprintf('%s (%s)', $feeds->title, $label);
```

aschempp commented on news-bundle/src/EventListener/DataContainer/PageListener.php:
```diff
@@ -0,0 +1,60 @@
+        $model = $this->framework->getAdapter(NewsArchiveModel::class);
+        $user = $this->security->getUser();
+
+        $archive = $user->isAdmin ? $model->findAll() : $model->findMultipleByIds($user->news);
```
```suggestion
        if (!$user instanceof BackendUser || $this->security->isGranted('ROLE_ADMIN')) {
            $archive = $model->findAll();
        } else {
            $archive = $model->findMultipleByIds($user->news);
        }
```
(might need `use Contao\BackendUser`)

aschempp commented on news-bundle/src/EventListener/DataContainer/PageListener.php:
```diff
@@ -0,0 +1,60 @@
+
+        $options = [];
+
+        if (null !== $archive) {
```
I think this whole method would be more performant by using real DB queries to only get title and ID instead of retrieving the full model in the back end.

aschempp commented on news-bundle/src/Resources/contao/dca/tl_layout.php:
```diff
@@ -22,7 +22,6 @@
 (
 	'exclude'         => true,
 	'inputType'       => 'checkbox',
-	'foreignKey'      => 'tl_news_feed.title',
```
could we use `tl_page.title` for this?

aschempp commented on news-bundle/src/Controller/NewsFeedController.php:
```diff
@@ -0,0 +1,257 @@
+            $this->responseTagger->addTags(['contao.db.news_archive.'.$archive]);
+        }
+
+        return $response;
```
should this also be tagged with `tl_page.ID`? or is that done automatically for pages @Toflar ?

Toflar replied:
No tagging is always the task of the respective controller. So it's done automatically for a page regular but not here :) But yes it should be tagged in order to remove the cache entry when you adjust the page alias etc. Should probably also add the jumpTo page ID of the news archive because if that changes, the URLs to the entries might change.

HOWEVER 😊 You are currently not caching anything. You only set the last modified header so the reverse proxy must revalidate and thus the tags are irrelevant ;) But we have cache settings in the page settings, why not applying them here too?

aschempp replied:
@bezin seems like this still applies?

bezin replied:
I think this was solved here: https://github.com/contao/contao/pull/4682/files#diff-e860c0628b064abb3cac479c7bafd78995fc7a8be5ef97dfc53f06580739fa69R84

aschempp commented on news-bundle/src/Resources/contao/languages/en/tl_page.xlf:
```diff
@@ -0,0 +1,55 @@
+      <trans-unit id="tl_page.feedDescription.1">
+        <source>Here you can enter a short description of the news feed.</source>
+      </trans-unit>
+      <trans-unit id="FEED_FORMAT.rss">
```
not sure this should to in `tl_page`. Also not sure it needs translation at all 😬

bezin replied:
Had the same two thoughts. I agree a translation is not required, but I would prefer decent labels ('RSS 2.0' rather than 'rss', JSON rather than 'json' and 'Atom' rather than 'atom') in several places. So the question is, where to put them then.

aschempp replied:
maybe we'll just hardcode them?

aschempp commented on news-bundle/src/Resources/contao/dca/tl_news_feed.php:
```diff
@@ -11,7 +11,6 @@
 use Contao\Automator;
 use Contao\Backend;
 use Contao\BackendUser;
-use Contao\CoreBundle\EventListener\Widget\HttpUrlListener;
```
shouldn't this whole file be dropped?

bezin replied:
Yes, will do this as last step (its one of the tasks) :-)

aschempp reviewed (changes requested):
Looks very promising! Added feedback and minor nitpicks, but overall its 🔥

fritzmg commented on news-bundle/src/Controller/NewsFeedController.php:
```diff
@@ -0,0 +1,257 @@
+        }
+
+        // TODO: Is this still necessary?
+        $origObjPage = $GLOBALS['objPage'] ?? null;
```
No, it is not necessary :). Now that this is a proper page controller, an appropriate `$GLOBALS['objPage']` within the correct website root will be available.

Original issue/PR: https://github.com/contao/contao/pull/2946

bezin replied:
Fixed in 2cdb25b

aschempp commented:
I also noticed FeedIO currently requires Guzzle (see https://github.com/alexdebril/feed-io/issues/398) 🙈

Toflar commented on news-bundle/src/Controller/NewsFeedController.php:
```diff
@@ -0,0 +1,281 @@
+     *
+     * @return Response The response object
+     */
+    private function setCacheHeaders(Response $response, PageModel $pageModel)
```
Wooooh, this looks very much like something we could use in other Contao controllers too! What do you think about moving that to `AbstractController`? :)

m-vo replied:
Or even the `EntityCacheTags` helper?

bezin replied:
I moved this method to `AbstractController` and extended from there

bezin commented:
> There are also some permissions for news feeds, i.e. which news feed a editor has access to and if he/she is allowed to create or delete them). Since this feed is now a page controller, access in general is handled by the page type limitation. The question is, if we want to keep this distinction?

What is your take on this? IMO we should drop these granular permissions and handle feeds like every other page, too.

aschempp commented:
> > There are also some permissions for news feeds, i.e. which news feed a editor has access to and if he/she is allowed to create or delete them). Since this feed is now a page controller, access in general is handled by the page type limitation. The question is, if we want to keep this distinction?
> 
> What is your take on this? IMO we should drop these granular permissions and handle feeds like every other page, too.

I think so too. Did we consider how this can be migrated on an update? Because I guess we cannot migrate the permissions?

bezin commented:
> I think so too. Did we consider how this can be migrated on an update? Because I guess we cannot migrate the permissions?

As there is no direct equivalent to each option after this gets merged I think we cannot migrate these. The baseline would be, that if an editor has the permission to at least one news feed, we should allow access to the new News Feed page type in general. However, this would mean the editor may now have permission to more feeds than previously.

bezin commented:
As you see I have added a `TransformFeedItem` to make the transformation from a news model to feed item exensible. However, I am not exactly sure about this and have some more cases, where I fail to see how to implement this the right way:

* I use feeds (custom implementation) with CleverReach to import Contao News into their editor. They do not wait long enough when requesting images for it work with our deferred image resizing. Hence I need a way to tell Contao to resize images right away in this case.

* For the same use case, I want to be able to overwrite image sizes in Image content elements to resize them to a size that is suitable for mailings. 

From https://github.com/contao/core-bundle/issues/1243:
* Add a namespace to the Feed object, e.g. for Facebook Instant Articles. The library we use supports this, but it must be extensible by a third party. 

Any comment much appreciated. Maybe the image resizing must be solved completely different anyways.

bezin commented:
I experimented with a provider service rather than an event (see https://github.com/bezin/contao/commit/19aad3be37fa3adfdfe4118b70c62ebe482479b1)

I implemented a FeedProvider that a) creates the feed instance, b) fetches the news articles depending on page settings and c) transforms articles to items. This way, you may decorate the provider and alter the feed instance and/or every item. 

Add additional namespaces to the feed instance:

```php
// src/App/NewsFeedProvider

class NewsFeedProvider implements FeedProviderInterface
{
  // ...
  public function getFeed(): FeedInterface
  {
    $feed = $this->inner->getFeed();
    $feed->addNS('content', 'http://purl.org/rss/1.0/modules/content/');
    return $feed;
  }
  // ...
}
```

Add a custom format options to `tl_page.feedSource` and alter item transformation all together in your decorator:

```php
// src/App/NewsFeedProvider

class NewsFeedProvider implements FeedProviderInterface
{
  // ...
  public function getItemFromModel(NewsModel $model): ItemInterface
  {
    if ('my_custom_source_option' !== $this->pageModel->feedSource) {
      return $this->inner->getItemFromModel($model);
    }
    
    // Handle item creation in your own way
  }
  // ...
}
```

What are your thoughts on this?

fritzmg commented:
Apparently you are not supposed to extend from the original service when decorating 😬 🤷 

I'd prefer an event still I think, since that is what we use for `/sitemap.xml` and `/robots.txt` as well.

m-vo commented:
> Apparently you are not supposed to extend from the original service when decorating

Yes, that's a bad idea in general, because you cannot guarantee that the original class will be in the correct state if you only pass on some calls. Implementing the interface, injecting the original service and using its public API is usually safer to do. :slightly_smiling_face: 

But tbh this is also true for every extended base class - so it primarily depends on if a class is designed for this use case or not (for non abstract classes: mostly not).

bezin commented:
> Apparently you are not supposed to extend from the original service when decorating 😬 🤷

Yes, you are right. This is actually wrong in the example – it is not required.

bezin commented:
I now implemented two events:
*  `FetchArticlesForFeed` (similiar to the `FetchNewsListItemHook`) to overwrite the logic to fetch the articles (filter by category, filter by author, filter by what not).
* `TransformArticleForFeed` to transform a news article to a feed item

The Contao\NewsBundle\EventListener\NewsFeedListener listens to both events. 

I still need to add tests and migrations.

aschempp commented on core-bundle/src/Resources/contao/models/PageModel.php:
```diff
@@ -111,6 +111,10 @@
+ * @property string         $newsArchives
+ * @property string         $feedFormat
+ * @property string         $feedSource
+ * @property integer        $maxFeedItems
```
not sure we're adding properties from extensions (`news-bundle`) in files from the core bundle @contao/developers ?

bezin replied:
Oh, yes you are probably right. Did not think about this. Can remove it, if required?

bezin replied:
We don't add anything to the other models either. Removed these in 0bd0c6361

aschempp commented on news-bundle/src/Controller/NewsFeedController.php:
```diff
@@ -0,0 +1,103 @@
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\NewsBundle\Controller;
```
I think we moved page controllers to a subfolder

```suggestion
namespace Contao\NewsBundle\Controller\Page;
```

aschempp commented on news-bundle/src/Controller/NewsFeedController.php:
```diff
@@ -0,0 +1,103 @@
+        ;
+
+        $event = new FetchArticlesForFeedEvent($feed, $request, $pageModel);
+        $this->container->get('event_dispatcher')->dispatch($event, FetchArticlesForFeedEvent::NAME);
```
If the event has only one purpose, I don't think it needs a name at all, since Symfony can register it on the class name (applies to the `TransformArticleForFeedEvent` as well)

bezin replied:
Fixed in fe0df19e2

aschempp commented on news-bundle/src/Controller/NewsFeedController.php:
```diff
@@ -0,0 +1,103 @@
+        $this->setCacheHeaders($response, $pageModel);
+
+        foreach (StringUtil::deserialize($pageModel->newsArchives, true) as $archive) {
+            $this->tagResponse('contao.db.tl_news_archive.'.$archive);
```
it feels a bit strange that this is done by the page controller, but actually fetching the feed items is not. The page controller can't really be sure if these archives are used. Shouldn't this rather be based on the articles that are fetched from the event (their PID)?

bezin replied:
Yes, you are right. That felt strange all along. I moved it up to the `foreach ($articles)` loop, where I add the tags for the article itself in 360608dca

aschempp commented on news-bundle/src/Controller/NewsFeedController.php:
```diff
@@ -0,0 +1,103 @@
+    public function configurePageRoute(PageRoute $route): void
+    {
+        $format = $route->getPageModel()->feedFormat;
+        $route->setUrlSuffix($this->urlSuffixes[$format]);
```
maybe handle if the `feedFormat` value in the DB is not found in the array?

bezin replied:
Fixed in d83623cbd

aschempp commented on news-bundle/src/Event/FetchArticlesForFeedEvent.php:
```diff
@@ -0,0 +1,54 @@
+{
+    public const NAME = 'contao_news.feed.fetch_articles';
+
+    private ?Collection $articles = null;
```
```suggestion
    private Collection|null $articles = null;
```

bezin replied:
Fixed  in 9455e015c

aschempp commented on news-bundle/src/Event/FetchArticlesForFeedEvent.php:
```diff
@@ -0,0 +1,54 @@
+        return $this->pageModel;
+    }
+
+    public function getArticles(): ?Collection
```
I think we should use an array of models here instead of a `Collection`, that would be more flexible to add own articles. And maybe a `addArticle` method as well?

bezin replied:
Fixed  in 9455e015c

aschempp commented on news-bundle/src/Event/TransformArticleForFeedEvent.php:
```diff
@@ -0,0 +1,66 @@
+{
+    public const NAME = 'contao_news.feed.transform_article';
+
+    private ?ItemInterface $item = null;
```
```suggestion
    private ItemInterface|null $item = null;
```

bezin replied:
Fixed  in 728791c05

aschempp commented on news-bundle/src/Event/TransformArticleForFeedEvent.php:
```diff
@@ -0,0 +1,66 @@
+        return $this->baseUrl;
+    }
+
+    public function getItem(): ?ItemInterface
```
```suggestion
    public function getItem(): ItemInterface|null
```

bezin replied:
Fixed in  728791c05

aschempp commented on news-bundle/src/EventListener/DataContainer/LayoutListener.php:
```diff
@@ -0,0 +1,48 @@
+use Contao\PageModel;
+use Symfony\Contracts\Translation\TranslatorInterface;
+
+class LayoutListener
```
I think this should be a `LayoutNewsfeedOptionsListener` or something like that 🙃

bezin replied:
Fixed in  5727555af

aschempp commented on news-bundle/src/EventListener/DataContainer/LayoutListener.php:
```diff
@@ -0,0 +1,48 @@
+    /**
+     * @Callback(table="tl_layout", target="fields.newsfeeds.options")
+     */
+    public function getAllFeeds(): array
```
```suggestion
    public function __invoke(): array
```

And the annotation should be moved to the class instead of the method

bezin replied:
Fixed in c65af05e4

aschempp commented on news-bundle/src/EventListener/DataContainer/LayoutListener.php:
```diff
@@ -0,0 +1,48 @@
+
+class LayoutListener
+{
+    public function __construct(private ContaoFramework $framework, private TranslatorInterface $translator)
```
`$translator` seems unused?

bezin replied:
Fixed

aschempp commented on news-bundle/src/EventListener/DataContainer/PageListener.php:
```diff
@@ -0,0 +1,68 @@
+        ;
+
+        if (!($user instanceof BackendUser || $this->security->isGranted('ROLE_ADMIN'))) {
+            $stmt->where('id', $user->news);
```
can we `->where()` on an array of IDs?

bezin replied:
Thanks for this comment – you are right of course. The implementation was actually completely bonkers. I fixed it and added basic unit tests in  718eab070

aschempp commented on news-bundle/src/EventListener/DataContainer/PageListener.php:
```diff
@@ -0,0 +1,68 @@
+
+        $options = [];
+
+        if ($results->rowCount() > 0) {
```
probably an unnecessary check, the foreach loop would just be empty?

aschempp commented on news-bundle/src/EventListener/GeneratePageListener.php:
```diff
@@ -43,21 +41,21 @@ public function __invoke(PageModel $pageModel, LayoutModel $layoutModel): void
+        $adapter = $this->framework->getAdapter(PageModel::class);
 
-        if (!($feeds = $adapter->findByIds($newsfeeds)) instanceof Collection) {
+        if (!($feeds = $adapter->findMultipleByIds($newsfeeds)) instanceof Collection) {
```
should we make sure these are feed pages?

bezin replied:
Fixed in 97f8d3dff

aschempp commented on news-bundle/src/EventListener/InsertTagsListener.php:
```diff
@@ -55,13 +56,17 @@ private function replaceNewsFeedInsertTag(string $feedId): string
     {
         $this->framework->initialize();
 
-        $adapter = $this->framework->getAdapter(NewsFeedModel::class);
```
I'm not sure we should actually keep the `news_feed` insert tag. It will not correctly work after an update, and it's the same as `link_url` anyway. Removing it (and maybe writing something to the logs) would probably be easier to spot errors.

bezin replied:
Yes, good point with `link_url` I'll trigger a `warning` in the logs.

bezin replied:
See 04a5351ae

aschempp commented on news-bundle/src/EventListener/NewsFeedListener.php:
```diff
@@ -0,0 +1,210 @@
+    public function onFetchArticlesForFeed(FetchArticlesForFeedEvent $event): void
+    {
+        $pageModel = $event->getPageModel();
+        $archives = StringUtil::deserialize($pageModel->newsArchives);
```
```suggestion
        $archives = StringUtil::deserialize($pageModel->newsArchives, true);
```

bezin replied:
Fixed in 395366e96

aschempp commented on news-bundle/src/EventListener/NewsFeedListener.php:
```diff
@@ -0,0 +1,210 @@
+        $newsModel = $this->framework->getAdapter(NewsModel::class);
+
+        $articles = $newsModel->findPublishedByPids($archives, $featured, $pageModel->maxFeedItems, 0, [
+            'return' => 'Collection',
```
```suggestion
            'return' => 'Array',
```
as suggested above

bezin replied:
Fixed in 395366e96

aschempp commented on news-bundle/src/EventListener/NewsFeedListener.php:
```diff
@@ -0,0 +1,210 @@
+
+        $this->pageModel = $event->getPageModel();
+        $this->request = $event->getRequest();
+        $this->baseUrl = $event->getBaseUrl();
```
this will introduce state to the listener service. I would pass on the event to the `getContent` instead

bezin replied:
Fixed in 395366e96

aschempp commented on news-bundle/src/EventListener/NewsFeedListener.php:
```diff
@@ -0,0 +1,210 @@
+        $event->setItem($item);
+    }
+
+    protected function getTitle(NewsModel $article, ItemInterface $item): string
```
any reason these methods are all `protected` instead of `private`?

bezin replied:
I experimented a bit with how to override the transformer in a listenr in your app. I ended up with extending the news bundle listener and overriding the individual `getTitle|Content|...` methods. This way one can override each field individually without having to overwrite everything else.

aschempp commented on news-bundle/src/Feed/FeedProvider.php:
```diff
@@ -0,0 +1,81 @@
+    {
+    }
+
+    public function setContext(PageModel $pageModel, Request $request, string $baseUrl): void
```
Without checking all the implications, doesn't the request already include the pageModel and baseUrl? Maybe passing the request to `getFeed` etc. would simplify the state problem?

bezin replied:
The Feed Provider is obsolete and will be deleted.

aschempp commented on news-bundle/src/Feed/FeedProvider.php:
```diff
@@ -0,0 +1,81 @@
+
+        return $newsModel->findPublishedByPids($archives, $featured, 0, 0, [
+            'return' => 'Collection',
+        ]);
```
Not sure what problem this is solving, but it could probably be simplified to this?

```suggestion
        return $newsModel->findPublishedByPids($archives, $featured, max((int) $this->pageModel->maxFeedItems, 0), 0, [
            'return' => 'Collection',
        ]);
```

bezin replied:
This code is obsolete since we only use events now.

aschempp commented on news-bundle/src/Feed/FeedProviderInterface.php:
```diff
@@ -0,0 +1,27 @@
+use FeedIo\FeedInterface;
+use Symfony\Component\HttpFoundation\Request;
+
+interface FeedProviderInterface
```
maybe I'm missing something, but can you explain what the `FeedProvider(Interface)` is for since we also have events?

bezin replied:
You are right, this code is obsolete, since we only use events now.

aschempp commented on news-bundle/src/Resources/contao/dca/tl_layout.php:
```diff
@@ -22,7 +22,7 @@
 (
 	'exclude'         => true,
 	'inputType'       => 'checkbox',
-	'foreignKey'      => 'tl_news_feed.title',
```
thinking about this, this should probably be a page picker instead? A checkbox list of pages does not make much sense 😆

bezin replied:
It is limited to only feed type pages in the corresponding listener, so it does not display too many pages.

aschempp commented on news-bundle/src/Resources/contao/languages/en/default.xlf:
```diff
@@ -38,6 +38,12 @@
+        <source>News Feed</source>
+      </trans-unit>
+      <trans-unit id="PTY.news_feed.1">
+        <source>Create a RSS/Atom/JSON Feed from your news archives</source>
```
```suggestion
        <source>Create a RSS/Atom/JSON feed from your news archives</source>
```

aschempp commented on news-bundle/src/Resources/contao/languages/en/modules.xlf:
```diff
@@ -8,9 +8,6 @@
       <trans-unit id="MOD.news.1">
         <source>Manage news and export them as RSS or Atom feed</source>
```
Does the RSS/Atom hint still belong here?

bezin replied:
Fixed in [8d472c1](https://github.com/contao/contao/pull/4682/commits/8d472c1fe5c35a578df5913375efd8fe2acb36e2)

aschempp reviewed (changes requested):
Love it ❤️
I've added a lot of minor nitpicks, but the overall concept totally makes sense to me and is a perfect example how page controllers can/should be used in the future 😎 

Regarding the icons, can a feed page really be visible or hidden from navigation? And what was changed on the `regular_2.svg` file?

aschempp commented on news-bundle/src/Controller/Page/NewsFeedController.php:
```diff
@@ -75,6 +75,7 @@ public function __invoke(Request $request, PageModel $pageModel): Response
             $this->container->get('event_dispatcher')->dispatch($event, TransformArticleForFeedEvent::NAME);
             $feed->add($event->getItem());
             $this->tagResponse($article);
+            $this->tagResponse('contao.db.tl_news_archive.'.$article->pid);
```
Not sure what happens if we tag the response with the same tag multiple times? @Toflar

bezin replied:
[They pass through `array_unique` every time you add tags. ](https://github.com/FriendsOfSymfony/FOSHttpCache/blob/master/src/ResponseTagger.php#L129)

bezin commented:
I cleaned up `regular_2.svg` but I reverted the change as it is unrelated. 

> Regarding the icons, can a feed page really be visible or hidden from navigation?

Besides the icons and more about the general functionality: I think the editor should be able to choose, if the page should be shown in the navigation. I mean after all, why should you not add a link to the feed in your main navigation?

I would also alyways exclude it from being indexed by the search engine. WDYT?

bezin committed:
Add news feed as page controller

aschempp commented on news-bundle/src/Controller/Page/NewsFeedController.php:
```diff
@@ -0,0 +1,103 @@
+        ;
+
+        $event = new FetchArticlesForFeedEvent($feed, $request, $pageModel);
+        $this->container->get('event_dispatcher')->dispatch($event);
```
we must also subscribe to the `event_dispatcher`

bezin replied:
The controller extends from `\Contao\CoreBundle\Controller\AbstractController` and that already subscribes to `event_dispatcher`

m-vo replied:
You still need to add it to your subscriptions. Never rely on the implementation detail of the base class.

aschempp commented on news-bundle/src/Controller/Page/NewsFeedController.php:
```diff
@@ -0,0 +1,103 @@
+        $this->container->get('event_dispatcher')->dispatch($event);
+
+        if (null === ($articles = $event->getArticles())) {
+            return new Response('', Response::HTTP_NO_CONTENT);
```
is it correct to render an empty response? Or should we render an empty feed (if feedio supports that)? Also, this should be a 200 not 204 status code.

> https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/204
>
> The HTTP 204 No Content success status response code indicates that a request has succeeded, but that the client doesn't need to navigate away from its current page.
> 
> This might be used, for example, when implementing "save and continue editing" functionality for a wiki site. > In this case a [PUT](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/PUT) request would be used to save the page, and the 204 No Content response would be sent to indicate that the editor should not be replaced by some other page.The HTTP 204 No Content success status response code indicates that a request has succeeded, but that the client doesn't need to navigate away from its current page.

bezin replied:
Yes, I agree. The feed itself exists, it just does not have any articles. So returning an empty feed seems right to me. Fixed in 699713dfe

bezin committed:
Improve GeneratePageListenerTest and NewsFeedListenerTest

aschempp commented on news-bundle/src/EventListener/DataContainer/PageListener.php:
```diff
@@ -0,0 +1,81 @@
+     */
+    public function onLoad(DataContainer $dc): void
+    {
+        $type = $this->connection->executeQuery('SELECT type FROM tl_page WHERE id = :id', ['id' => $dc->id])->fetchFirstColumn();
```
```suggestion
        $type = $dc->getCurrentRecord()['type'];
```

bezin replied:
Implemented in 9d1929cd2

aschempp commented on news-bundle/src/Migration/FeedMigration.php:
```diff
@@ -0,0 +1,107 @@
+
+        $result = $this->connection->executeQuery('SELECT COUNT(id) AS count FROM tl_news_feed')->fetchFirstColumn();
+
+        return !empty($result) && $result[0] > 0;
```
```suggestion
        return $this->connection->fetchOne('SELECT COUNT(id)  FROM tl_news_feed') > 0;
```

bezin replied:
Fixed in a62be998d

aschempp commented on news-bundle/src/Migration/FeedMigration.php:
```diff
@@ -0,0 +1,107 @@
+                    tl_page
+                ADD
+                    $field $definition
+            ");
```
Either the `shouldRun` needs to make sure the migration is only executed if these fields don't exist, or better we should probably check for the fields here and skip existing ones.

bezin replied:
Fixed in ccc498110

aschempp commented on news-bundle/src/Migration/FeedMigration.php:
```diff
@@ -0,0 +1,107 @@
+        // Find first root page, if none matches by dns and language
+        if (!$page) {
+            $page = $this->connection
+                ->executeQuery("SELECT id FROM tl_page WHERE type = 'root' ORDER BY sorting ASC LIMIT 1")
```
```suggestion
                ->executeQuery("SELECT id FROM tl_page WHERE type = 'root' AND fallback=1 ORDER BY sorting ASC LIMIT 1")
```

bezin replied:
Fixed in ccc498110

aschempp commented on news-bundle/src/Migration/FeedMigration.php:
```diff
@@ -0,0 +1,107 @@
+            ;
+        }
+
+        return $page[0] ?? null;
```
could also use `->connection->fetchOne()`

bezin replied:
Fixed in a62be998d

bezin committed:
Fix issues caused by static classes

bezin committed:
Update PageListener

bezin committed:
Update FeedMigration

bezin committed:
Rework FeedMigration

bezin committed:
cs

bezin committed:
Return empty feed instead of empty response, if no articles where found

fritzmg committed:
fix news feed migration

fritzmg reviewed (commented):
I found a bug in the FeedMigration - see PR: https://github.com/bezin/contao/pull/5

fritzmg committed:
fix various small issues

bezin committed:
cs

bezin committed:
Fix tl_news_archive dca

bezin committed:
fix news-bundle/composer.json

bezin committed:
Move feed icon to new location in /contao

bezin committed:
Move dca and language files to new location in /contao

ausi commented:
> Add tests

@bezin Is this still planned? 😬

bezin commented:
Tests are there, the check was missing

leofeyer committed:
Merge branch '5.0' into news-feed-as-controller

leofeyer committed:
CS

leofeyer commented on core-bundle/src/Controller/AbstractController.php:
```diff
@@ -72,4 +74,53 @@ protected function getCsrfFormOptions(): array
+    protected function setCacheHeaders(Response $response, PageModel $pageModel)
+    {
+        // Do not cache the response if caching was not configured at all or disabled explicitly
+        if ((false === $pageModel->cache || $pageModel->cache < 1) && (false === $pageModel->clientCache || $pageModel->clientCache < 1)) {
```
```suggestion
        if ($pageModel->cache < 1 && $pageModel->clientCache < 1)
```

bezin replied:
Did you use $objPage by accident?

leofeyer replied:
Yes, sorry. I have adjusted the suggestion.

bezin replied:
Fixed in f57ef3ab5

leofeyer commented on core-bundle/src/Controller/AbstractController.php:
```diff
@@ -72,4 +74,53 @@ protected function getCsrfFormOptions(): array
+            }
+
+            // Tag the page (see #2137)
+            $this->tagResponse($pageModel);
```
```suggestion
            $this->container->get('contao.cache.entity_tags')->tagWithModelInstance($pageModel);
```

bezin replied:
Changed in 0a0fdf4e4

leofeyer replied:
The change is wrong, I‘m afraid. Why did you not commit my suggestion?

bezin replied:
Fixed in 0181a84c8

leofeyer reviewed (changes requested):
Setting `cache` or `clientCache` to `false` is no longer supported in Contao 5.

leofeyer commented on news-bundle/config/services.yaml:
```diff
@@ -30,9 +48,22 @@ services:
+            - '@contao.cache.entity_tags'
+        tags:
+            - { name: kernel.event_listener, event: Contao\NewsBundle\Event\FetchArticlesForFeedEvent, method: onFetchArticlesForFeed }
+            - { name: kernel.event_listener, event: Contao\NewsBundle\Event\TransformArticleForFeedEvent, method: onTransformArticleForFeed }
```
Please use PHP 8 attributes to tag the listener.

bezin replied:
Fixed in d8ddd2965

leofeyer commented on news-bundle/contao/classes/News.php:
```diff
@@ -26,261 +21,6 @@ class News extends Frontend
-	 *
-	 * @param integer $intId
-	 */
-	public function generateFeed($intId)
```
@contao/developers Do we really want to remove all these methods? None of them are deprecated in Contao 4.13.

Toflar replied:
It's not something we can deprecate. Deprecating it would mean we have to write "do this instead". But here is no "do this instead" because there is no alternative. In Contao 4.13 the data was written to disk using this method. Now it's a controller. It's an entirely different concept so what you'll have to do depends on what you did with this method.

leofeyer commented on news-bundle/src/Controller/Page/NewsFeedController.php:
```diff
@@ -0,0 +1,101 @@
+
+/**
+ * @Page(contentComposition=false)
+ */
```
```suggestion
#[AsPage(contentComposition: false)]
```

You also have to add a `use` statement for the class.

bezin replied:
Fixed in d8ddd2965

leofeyer commented on news-bundle/src/Controller/Page/NewsFeedController.php:
```diff
@@ -0,0 +1,101 @@
+        if (null !== ($articles = $event->getArticles())) {
+            foreach ($articles as $article) {
+                $event = new TransformArticleForFeedEvent($article, $feed, $pageModel, $request, $baseUrl);
+                $this->container->get('event_dispatcher')->dispatch($event);
```
This should not be inside a loop.

bezin replied:
Fixed in 3d6ac3e04

leofeyer commented on news-bundle/src/EventListener/DataContainer/LayoutFeedOptionsListener.php:
```diff
@@ -0,0 +1,47 @@
+
+/**
+ * @Callback(table="tl_layout", target="fields.newsfeeds.options")
+ */
```
This should be a PHP 8 attribute as well.

bezin replied:
Fixed in d8ddd2965

leofeyer commented on news-bundle/src/EventListener/DataContainer/PageListener.php:
```diff
@@ -0,0 +1,81 @@
+
+    /**
+     * @Callback(table="tl_page", target="config.onload")
+     */
```
All of these should be PHP 8 attributes.

bezin replied:
Fixed in d8ddd2965

leofeyer commented on news-bundle/src/EventListener/GeneratePageListener.php:
```diff
@@ -43,21 +41,24 @@ public function __invoke(PageModel $pageModel, LayoutModel $layoutModel): void
+            if (NewsFeedController::TYPE !== $feed->type) {
+                continue;
+            }
+            // TODO: Use ResponseContext, once it supports appending to <head>
```
@Toflar I figured it already would?

Toflar replied:
It does not, no.

leofeyer commented on news-bundle/src/EventListener/NewsFeedListener.php:
```diff
@@ -0,0 +1,213 @@
+        $event->setItem($item);
+    }
+
+    protected function getTitle(NewsModel $article, ItemInterface $item, TransformArticleForFeedEvent $event): string
```
Why is it necessary to pass the `$item` and `$event` objects of none of them are used?

bezin replied:
Because it is `protected` and you may extend this class from your own listener to overwrite the methods.

ausi replied:
Since our event listeners are not part of the BC promise I’d go with `private` methods here.

leofeyer replied:
Me too. Extending listeners is kind of an anti-pattern.

bezin replied:
Agreed, however it saves you work when you only want to extend one or some methods. 
But in lack of a better idea and the fact that I just want to get this across the finish line, I made them all private in 9f9f381a7

bezin replied:
Removed unused arguments in [8344a47](https://github.com/contao/contao/pull/4682/commits/8344a47a174d419549371e79c7556a5f3267bfb5)

leofeyer commented on news-bundle/tests/Controller/Page/NewsFeedControllerTest.php:
```diff
@@ -0,0 +1,187 @@
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Controller\Page;
```
This seems wrong. It should probably be `Contao\NewsBundle\Tests\Controller\Page`.

bezin replied:
Fixed in a8cda0c9c

leofeyer commented on news-bundle/tests/EventListener/DataContainer/LayoutFeedOptionsListenerTest.php:
```diff
@@ -0,0 +1,79 @@
+ * @license LGPL-3.0-or-later
+ */
+
+namespace EventListener\DataContainer;
```
This seems wrong, too.

bezin replied:
Fixed in a8cda0c9c

leofeyer commented:
Is it really necessary to duplicate the dummy images?

bezin committed:
Do not get event dispatcher in articles loop

bezin committed:
Fix namespaces

bezin committed:
Use PHP 8 attribtues

bezin committed:
Use tagWithModelInstance directly

bezin commented:
> Is it really necessary to duplicate the dummy images?

Since I need to check if the files exist, I would say yes. I do not see how I could split up the `getEnclosure` method in `NewsFeedListener` in a meaningful way.

bezin committed:
Update AbstractController

ausi commented:
> > Is it really necessary to duplicate the dummy images?
> 
> Since I need to check if the files exist, I would say yes.

Couldn’t you create those images on the fly when running the tests? like we do here: https://github.com/contao/contao/blob/5480dd09fd77c2ad5fdea497dddd932986f9bdf9/core-bundle/tests/Image/Preview/PreviewFactoryTest.php#L348-L351

leofeyer committed:
Merge branch '5.0' into news-feed-as-controller

leofeyer committed:
CS

leofeyer commented on news-bundle/src/Controller/Page/NewsFeedController.php:
```diff
@@ -0,0 +1,104 @@
+        return $response;
+    }
+
+    public function configurePageRoute(PageRoute $route): void
```
Why is this method public?

bezin replied:
See [DynamicRouteInterface](https://github.com/contao/contao/blob/a5c401df53881cd09edd0e5dfc9813dbb659c251/core-bundle/src/Routing/Page/DynamicRouteInterface.php)

leofeyer commented on news-bundle/src/Controller/Page/NewsFeedController.php:
```diff
@@ -0,0 +1,104 @@
+        $route->setUrlSuffix($this->urlSuffixes[$format]);
+    }
+
+    public function getUrlSuffixes(): array
```
What's the purpose of this method?

bezin replied:
See [DynamicRouteInterface](https://github.com/contao/contao/blob/a5c401df53881cd09edd0e5dfc9813dbb659c251/core-bundle/src/Routing/Page/DynamicRouteInterface.php)

leofeyer committed:
CS

leofeyer committed:
CS

Toflar commented on core-bundle/src/Controller/AbstractController.php:
```diff
@@ -72,4 +74,53 @@ protected function getCsrfFormOptions(): array
+            }
+
+            // Tag the page (see #2137)
+            $this->container->get('contao.cache.entity_tags')->tagWithModelInstance($tags);
```
What's `$tags` here? :)

bezin replied:
Fixed in 0181a84

bezin committed:
Refactor NewsFeedListenerTest to create images on the fly

bezin committed:
Fix AbstractController

bezin commented:
> Couldn’t you create those images on the fly when running the tests? like we do here:

Did so in [931f26b](https://github.com/contao/contao/pull/4682/commits/931f26b2b454981ae5c580645ab9548cc500239e) Thanks for the hint!

bezin committed:
Make protected methods in NewsFeedListener private

bezin committed:
cs

bezin committed:
Remove unnecessary function arguments

leofeyer committed:
CS

leofeyer committed:
Adjust the palette and the language strings

leofeyer committed:
Use a separate column to store the feed description

leofeyer committed:
Add `COLLATE ascii_bin` where applicable

leofeyer commented:
Thank you @bezin.
